### PR TITLE
Development/duke

### PR DESCRIPTION
**Pull request by jodi-one**
Please review.

### CODE REVIEW

File: jodi_odi12/src/main/groovy/findPackagesLastModified.groovy [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import bsh.StringUtil
+import jnr.ffi.Struct
+import oracle.odi.core.OdiInstance
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findPackagesLastModified.groovy:3]:
```suggestion
```

File: jodi_odi12/src/main/groovy/findPackagesLastModified.groovy [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import oracle.odi.core.persistence.IOdiEntityManager
+import oracle.odi.core.persistence.transaction.ITransactionManager
+import oracle.odi.core.persistence.transaction.ITransactionStatus
+import oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition
+import oracle.odi.domain.project.OdiFolder
+import oracle.odi.domain.project.OdiPackage
+import oracle.odi.domain.project.finder.IOdiPackageFinder
+
+import java.util.Collection
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findPackagesLastModified.groovy:12]:
```suggestion
```

File: jodi_odi12/src/main/groovy/findPackagesLastModified.groovy [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+int lastXresults = 10;
+
+oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition txnDef =   new DefaultTransactionDefinition()
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findPackagesLastModified.groovy:16]:
```suggestion
DefaultTransactionDefinition txnDef =   new DefaultTransactionDefinition()
```

File: jodi_odi12/src/main/groovy/findPackagesLastModified.groovy [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ITransactionManager tm = odiInstance.getTransactionManager()
+IOdiEntityManager tem = odiInstance.getTransactionalEntityManager()
+ITransactionStatus txnStatus = tm.getTransaction(txnDef)
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findPackagesLastModified.groovy:19]:
```suggestion
tm.getTransaction(txnDef)
```

File: jodi_odi12/src/main/groovy/findPackagesLastModified.groovy [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+IOdiPackageFinder finder = (IOdiPackageFinder) tem.getFinder(OdiPackage.class)
+
+java.util.Collection<OdiPackage> packages = finder.findAll()
+
+Collections.sort(packages, new Comparator<OdiPackage>() {
+    @Override
+    public int compare(OdiPackage p1, OdiPackage p2) {
+        return p2.getLastDate().compareTo(p1.getLastDate());
+    }
+});
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findPackagesLastModified.groovy:29]:
```suggestion
Collection<OdiPackage> packages = finder.findAll().sort { p1, p2 -> p2.getLastDate().compareTo(p1.getLastDate()) }
```

File: jodi_odi12/src/main/groovy/findPackagesLastModified.groovy [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+def getFolderPath( OdiFolder folder, String base ){
+    if(folder.getParentFolder() == null){
+        return  "/" + folder.getName() + base
+    }else {
+        return getFolderPath(folder.getParentFolder() , base + "/" + folder.getName() )
+    }
+
+}
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findPackagesLastModified.groovy:38]:
```suggestion
def getFolderPath(OdiFolder folder, String base) {
   def separator = FileSystems.getDefault().getSeparator()
   if (folder.getParentFolder() == null) {
      return separator + folder.getName() + base
   }
   return getFolderPath(folder.getParentFolder(), base + separator + folder.getName())
}
```

File: jodi_odi12/src/main/groovy/findProcedures.groovy [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import bsh.StringUtil
+import jnr.ffi.Struct
+import oracle.odi.core.OdiInstance
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findProcedures.groovy:3]:
```suggestion
```

File: jodi_odi12/src/main/groovy/findProcedures.groovy [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import oracle.odi.core.persistence.IOdiEntityManager
+import oracle.odi.core.persistence.transaction.ITransactionManager
+import oracle.odi.core.persistence.transaction.ITransactionStatus
+import oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition
+import oracle.odi.domain.project.OdiFolder
+import oracle.odi.domain.project.OdiUserProcedure
+import oracle.odi.domain.project.finder.IOdiUserProcedureFinder
+
+import java.util.Collection
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findProcedures.groovy:12]:
```suggestion
```

File: jodi_odi12/src/main/groovy/findProcedures.groovy [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+
+oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition txnDef =   new DefaultTransactionDefinition()
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findProcedures.groovy:15]:
```suggestion
DefaultTransactionDefinition txnDef =   new DefaultTransactionDefinition()
```

File: jodi_odi12/src/main/groovy/findProcedures.groovy [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+ITransactionManager tm = odiInstance.getTransactionManager()
+IOdiEntityManager tem = odiInstance.getTransactionalEntityManager()
+ITransactionStatus txnStatus = tm.getTransaction(txnDef)
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findProcedures.groovy:18]:
```suggestion
tm.getTransaction(txnDef)
```

File: jodi_odi12/src/main/groovy/findProcedures.groovy [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+IOdiUserProcedureFinder finder = (IOdiUserProcedureFinder) tem.getFinder(OdiUserProcedure.class)
+
+FolderPN = "Folder"
+SEARCH = "Search"
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findProcedures.groovy:22]:
```suggestion
def SEARCH = "Search"
```

File: jodi_odi12/src/main/groovy/findProcedures.groovy [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+java.util.Collection<OdiUserProcedure> procedures = finder.findAll()
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findProcedures.groovy:24]:
```suggestion
Collection<OdiUserProcedure> procedures = finder.findAll()
```

File: jodi_odi12/src/main/groovy/findProcedures.groovy [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+def getFolderPath(OdiFolder folder, String base ){
+    if(folder.getParentFolder() == null){
+        return  "/" + folder.getName() + base
+    }else {
+        return getFolderPath(folder.getParentFolder() , base + "/" + folder.getName() )
+    }
+
+}
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findProcedures.groovy:33]:
```suggestion
def getFolderPath(OdiFolder folder, String base) {
   def separator = FileSystems.getDefault().getSeparator()
   if (folder.getParentFolder() == null) {
      return separator + folder.getName() + base
   }
   return getFolderPath(folder.getParentFolder(), base + separator + folder.getName())
}
```

File: jodi_odi12/src/main/groovy/findProceduresLastModified.groovy [hunk 14]
```diff
@@ -0,0 +1,41 @@
+import bsh.StringUtil
+import jnr.ffi.Struct
+import oracle.odi.core.OdiInstance
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findProceduresLastModified.groovy:3]:
```suggestion
```

File: jodi_odi12/src/main/groovy/findProceduresLastModified.groovy [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+import oracle.odi.core.persistence.IOdiEntityManager
+import oracle.odi.core.persistence.transaction.ITransactionManager
+import oracle.odi.core.persistence.transaction.ITransactionStatus
+import oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition
+import oracle.odi.domain.project.OdiFolder
+import oracle.odi.domain.project.OdiUserProcedure
+import oracle.odi.domain.project.finder.IOdiUserProcedureFinder
+
+import java.util.Collection
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findProceduresLastModified.groovy:12]:
```suggestion
```

File: jodi_odi12/src/main/groovy/findProceduresLastModified.groovy [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+int lastXresults = 10;
+
+oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition txnDef =   new DefaultTransactionDefinition()
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findProceduresLastModified.groovy:16]:
```suggestion
DefaultTransactionDefinition txnDef = new DefaultTransactionDefinition()
```

File: jodi_odi12/src/main/groovy/findProceduresLastModified.groovy [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+ITransactionManager tm = odiInstance.getTransactionManager()
+IOdiEntityManager tem = odiInstance.getTransactionalEntityManager()
+ITransactionStatus txnStatus = tm.getTransaction(txnDef)
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findProceduresLastModified.groovy:19]:
```suggestion
tm.getTransaction(txnDef)
```

File: jodi_odi12/src/main/groovy/findProceduresLastModified.groovy [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+IOdiUserProcedureFinder finder = (IOdiUserProcedureFinder) tem.getFinder(OdiUserProcedure.class)
+
+java.util.Collection<OdiUserProcedure> procedures = finder.findAll()
+
+Collections.sort(procedures, new Comparator<OdiUserProcedure>() {
+    @Override
+    public int compare(OdiUserProcedure m1, OdiUserProcedure m2) {
+        return m2.getLastDate().compareTo(m1.getLastDate());
+    }
+});
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findProceduresLastModified.groovy:29]:
```suggestion
Collection<OdiUserProcedure> procedures = finder.findAll().sort { p1, p2 -> p2.getLastDate().compareTo(p1.getLastDate()) }
```

File: jodi_odi12/src/main/groovy/findProceduresLastModified.groovy [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+def getFolderPath( OdiFolder folder, String base ){
+    if(folder.getParentFolder() == null){
+        return  "/" + folder.getName() + base
+    }else {
+        return getFolderPath(folder.getParentFolder() , base + "/" + folder.getName() )
+    }
+
+}
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findProceduresLastModified.groovy:38]:
```suggestion
def getFolderPath(OdiFolder folder, String base) {
   def separator = FileSystems.getDefault().getSeparator()
   if (folder.getParentFolder() == null) {
      return separator + folder.getName() + base
   }
   return getFolderPath(folder.getParentFolder(), base + separator + folder.getName())
}
```

File: jodi_odi12/src/main/groovy/findScenarios.groovy [hunk 20]
```diff
@@ -0,0 +1,44 @@
+import bsh.StringUtil
+import jnr.ffi.Struct
+import oracle.odi.core.OdiInstance
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findScenarios.groovy:3]:
```suggestion
```

File: jodi_odi12/src/main/groovy/findScenarios.groovy [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+import oracle.odi.core.persistence.IOdiEntityManager
+import oracle.odi.core.persistence.transaction.ITransactionManager
+import oracle.odi.core.persistence.transaction.ITransactionStatus
+import oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition
+import oracle.odi.domain.runtime.loadplan.OdiLoadPlan
+import oracle.odi.domain.runtime.scenario.OdiScenario
+import oracle.odi.domain.runtime.scenario.OdiScenarioFolder
+import oracle.odi.domain.runtime.scenario.finder.IOdiScenarioFinder
+
+import java.util.Collection
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findScenarios.groovy:13]:
```suggestion
```

File: jodi_odi12/src/main/groovy/findScenarios.groovy [hunk 22]
```diff
@@ -0,0 +1,44 @@
+import bsh.StringUtil
+import jnr.ffi.Struct
+import oracle.odi.core.OdiInstance
+import oracle.odi.core.persistence.IOdiEntityManager
+import oracle.odi.core.persistence.transaction.ITransactionManager
+import oracle.odi.core.persistence.transaction.ITransactionStatus
+import oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition
+import oracle.odi.domain.runtime.loadplan.OdiLoadPlan
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findScenarios.groovy:8]:
```suggestion
```

File: jodi_odi12/src/main/groovy/findScenarios.groovy [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+import oracle.odi.domain.runtime.scenario.OdiScenarioFolder
+import oracle.odi.domain.runtime.scenario.finder.IOdiScenarioFinder
+
+import java.util.Collection
+
+
+oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition txnDef =   new DefaultTransactionDefinition()
+ITransactionManager tm = odiInstance.getTransactionManager()
+IOdiEntityManager tem = odiInstance.getTransactionalEntityManager()
+ITransactionStatus txnStatus = tm.getTransaction(txnDef)
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findScenarios.groovy:19]:
```suggestion
tm.getTransaction(txnDef)
```

File: jodi_odi12/src/main/groovy/findScenarios.groovy [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+IOdiScenarioFinder finder = (IOdiScenarioFinder) tem.getFinder(OdiScenario.class)
+
+FolderPN = "Folder"
+SEARCH = "Search"
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findScenarios.groovy:23]:
```suggestion
def SEARCH = "Search"
```

File: jodi_odi12/src/main/groovy/findScenarios.groovy [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+
+java.util.Collection<OdiScenario> scenarios = finder.findAll()
+
+Collections.sort(scenarios, new Comparator<OdiScenario>() {
+    @Override
+    public int compare(OdiScenario s1, OdiScenario s2) {
+        return s2.getLastDate().compareTo(s1.getLastDate());
+    }
+});
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findScenarios.groovy:32]:
```suggestion
Collection<OdiScenario> scenarios = finder.findAll().sort { s1, s2 -> s2.getLastDate().compareTo(s1.getLastDate()) }
```

File: jodi_odi12/src/main/groovy/findScenarios.groovy [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+
+def getFolderPath(OdiScenarioFolder folder, String base ){
+    if(folder.getParentScenFolder() == null){
+        return  "/" + folder.getName() + base
+    }else {
+        return getFolderPath(folder.getParentScenFolder() , base + "/" + folder.getName() )
+    }
+
+}
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findScenarios.groovy:41]:
```suggestion
def getFolderPath(OdiScenarioFolder folder, String base) {
   def separator = FileSystems.getDefault().getSeparator()
   if (folder.getParentScenFolder() == null) {
      return separator + folder.getName() + base
   }
   return getFolderPath(folder.getParentScenFolder(), base + separator + folder.getName())
}
```

File: jodi_odi12/src/main/groovy/findScenariosLastModified.groovy [hunk 27]
```diff
@@ -0,0 +1,41 @@
+import bsh.StringUtil
+import jnr.ffi.Struct
+import oracle.odi.core.OdiInstance
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findScenariosLastModified.groovy:3]:
```suggestion
```

File: jodi_odi12/src/main/groovy/findScenariosLastModified.groovy [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+import oracle.odi.core.persistence.IOdiEntityManager
+import oracle.odi.core.persistence.transaction.ITransactionManager
+import oracle.odi.core.persistence.transaction.ITransactionStatus
+import oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition
+import oracle.odi.domain.runtime.scenario.OdiScenario
+import oracle.odi.domain.runtime.scenario.OdiScenarioFolder
+import oracle.odi.domain.runtime.scenario.finder.IOdiScenarioFinder
+
+import java.util.Collection
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findScenariosLastModified.groovy:12]:
```suggestion
```

File: jodi_odi12/src/main/groovy/findScenariosLastModified.groovy [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+
+int lastXresults = 10;
+
+oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition txnDef =   new DefaultTransactionDefinition()
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findScenariosLastModified.groovy:16]:
```suggestion
DefaultTransactionDefinition txnDef = new DefaultTransactionDefinition()
```

File: jodi_odi12/src/main/groovy/findScenariosLastModified.groovy [hunk 30]
```diff
@@ -0,0 +1,41 @@
...
+ITransactionManager tm = odiInstance.getTransactionManager()
+IOdiEntityManager tem = odiInstance.getTransactionalEntityManager()
+ITransactionStatus txnStatus = tm.getTransaction(txnDef)
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findScenariosLastModified.groovy:19]:
```suggestion
tm.getTransaction(txnDef)
```

File: jodi_odi12/src/main/groovy/findScenariosLastModified.groovy [hunk 31]
```diff
@@ -0,0 +1,41 @@
...
+IOdiScenarioFinder finder = (IOdiScenarioFinder) tem.getFinder(OdiScenario.class)
+
+java.util.Collection<OdiScenario> scenarios = finder.findAll()
+
+Collections.sort(scenarios, new Comparator<OdiScenario>() {
+    @Override
+    public int compare(OdiScenario s1, OdiScenario s2) {
+        return s2.getLastDate().compareTo(s1.getLastDate());
+    }
+});
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findScenariosLastModified.groovy:29]:
```suggestion
Collection<OdiScenario> scenarios = finder.findAll().sort { s1, s2 -> s2.getLastDate().compareTo(s1.getLastDate()) }
```

File: jodi_odi12/src/main/groovy/findScenariosLastModified.groovy [hunk 32]
```diff
@@ -0,0 +1,41 @@
...
+
+def getFolderPath(OdiScenarioFolder folder, String base ){
+    if(folder.getParentScenFolder() == null){
+        return  "/" + folder.getName() + base
+    }else {
+        return getFolderPath(folder.getParentScenFolder() , base + "/" + folder.getName() )
+    }
+
+}
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findScenariosLastModified.groovy:38]:
```suggestion
def getFolderPath(OdiScenarioFolder folder, String base) {
   def separator = FileSystems.getDefault().getSeparator()
   if (folder.getParentScenFolder() == null) {
      return separator + folder.getName() + base
   }
   return getFolderPath(folder.getParentScenFolder(), base + separator + folder.getName())
}
```

File: jodi_odi12/src/main/groovy/findLoadPlan.groovy [hunk 33]
```diff
@@ -0,0 +1,44 @@
...
+
+
+oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition txnDef =   new DefaultTransactionDefinition()
+ITransactionManager tm = odiInstance.getTransactionManager()
+IOdiEntityManager tem = odiInstance.getTransactionalEntityManager()
+ITransactionStatus txnStatus = tm.getTransaction(txnDef)
+IOdiLoadPlanFinder finder = (IOdiLoadPlanFinder) tem.getFinder(OdiLoadPlan.class)
+
+FolderPN = "Folder"
+SEARCH = "Search"
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findLoadPlan.groovy:22]: Constant value, should be declared with `def` (like `final` in java)
```suggestion
def SEARCH = "Search"
```

File: jodi_odi12/src/main/groovy/findMappings.groovy [hunk 34]
```diff
@@ -0,0 +1,36 @@
...
+import oracle.odi.core.persistence.transaction.ITransactionStatus
+import oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition
+import oracle.odi.domain.adapter.project.IMapping
+import oracle.odi.domain.mapping.Mapping
+import oracle.odi.domain.mapping.finder.IMappingFinder
+import oracle.odi.domain.project.OdiFolder
+
+import java.util.Collection
+
+
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findMappings.groovy:15]: Used later in suggestion...
```suggestion
import java.nio.file.FileSystems
```

File: jodi_odi12/src/main/groovy/findPackages.groovy [hunk 35]
```diff
@@ -0,0 +1,35 @@
...
+import oracle.odi.core.persistence.IOdiEntityManager
+import oracle.odi.core.persistence.transaction.ITransactionManager
+import oracle.odi.core.persistence.transaction.ITransactionStatus
+import oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition
+import oracle.odi.domain.project.OdiFolder
+import oracle.odi.domain.project.OdiPackage
+import oracle.odi.domain.project.finder.IOdiPackageFinder
+
+import java.util.Collection
+
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findPackages.groovy:13]: Used later in suggestion...
```suggestion
import java.nio.file.FileSystems
```

File: jodi_odi12/src/main/groovy/findPackagesLastModified.groovy [hunk 36]
```diff
@@ -0,0 +1,41 @@
...
+import oracle.odi.core.persistence.IOdiEntityManager
+import oracle.odi.core.persistence.transaction.ITransactionManager
+import oracle.odi.core.persistence.transaction.ITransactionStatus
+import oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition
+import oracle.odi.domain.project.OdiFolder
+import oracle.odi.domain.project.OdiPackage
+import oracle.odi.domain.project.finder.IOdiPackageFinder
+
+import java.util.Collection
+
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findPackagesLastModified.groovy:13]: Used later in suggestion...
```suggestion
import java.nio.file.FileSystems

```

File: jodi_odi12/src/main/groovy/findProcedures.groovy [hunk 37]
```diff
@@ -0,0 +1,35 @@
...
+import oracle.odi.core.persistence.IOdiEntityManager
+import oracle.odi.core.persistence.transaction.ITransactionManager
+import oracle.odi.core.persistence.transaction.ITransactionStatus
+import oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition
+import oracle.odi.domain.project.OdiFolder
+import oracle.odi.domain.project.OdiUserProcedure
+import oracle.odi.domain.project.finder.IOdiUserProcedureFinder
+
+import java.util.Collection
+
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findProcedures.groovy:13]: Used later in suggestion...
```suggestion
import java.nio.file.FileSystems
```

File: jodi_odi12/src/main/groovy/findProceduresLastModified.groovy [hunk 38]
```diff
@@ -0,0 +1,41 @@
...
+import oracle.odi.core.persistence.IOdiEntityManager
+import oracle.odi.core.persistence.transaction.ITransactionManager
+import oracle.odi.core.persistence.transaction.ITransactionStatus
+import oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition
+import oracle.odi.domain.project.OdiFolder
+import oracle.odi.domain.project.OdiUserProcedure
+import oracle.odi.domain.project.finder.IOdiUserProcedureFinder
+
+import java.util.Collection
+
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findProceduresLastModified.groovy:13]: Used later in suggestion...
```suggestion
import java.nio.file.FileSystems
```

File: jodi_odi12/src/main/groovy/findScenarios.groovy [hunk 39]
```diff
@@ -0,0 +1,44 @@
...
+import oracle.odi.core.persistence.transaction.ITransactionManager
+import oracle.odi.core.persistence.transaction.ITransactionStatus
+import oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition
+import oracle.odi.domain.runtime.loadplan.OdiLoadPlan
+import oracle.odi.domain.runtime.scenario.OdiScenario
+import oracle.odi.domain.runtime.scenario.OdiScenarioFolder
+import oracle.odi.domain.runtime.scenario.finder.IOdiScenarioFinder
+
+import java.util.Collection
+
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findScenarios.groovy:14]: Used later in suggestion...
```suggestion
import java.nio.file.FileSystems
```

File: jodi_odi12/src/main/groovy/findScenariosLastModified.groovy [hunk 40]
```diff
@@ -0,0 +1,41 @@
...
+import oracle.odi.core.persistence.IOdiEntityManager
+import oracle.odi.core.persistence.transaction.ITransactionManager
+import oracle.odi.core.persistence.transaction.ITransactionStatus
+import oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition
+import oracle.odi.domain.runtime.scenario.OdiScenario
+import oracle.odi.domain.runtime.scenario.OdiScenarioFolder
+import oracle.odi.domain.runtime.scenario.finder.IOdiScenarioFinder
+
+import java.util.Collection
+
```
Comment (eelse) [inline jodi_odi12/src/main/groovy/findScenariosLastModified.groovy:13]: Used later in suggestion...
```suggestion
import java.nio.file.FileSystems
```

**Review by eelse — CHANGES_REQUESTED** [review]

File: jodi_odi12/src/main/groovy/findLoadPlansLastModified.groovy [hunk 41]
```diff
@@ -0,0 +1,41 @@
...
+import oracle.odi.domain.runtime.scenario.OdiScenarioFolder
+
+import java.util.Collection
+
+int lastXresults = 10;
+
+oracle.odi.core.persistence.transaction.support.DefaultTransactionDefinition txnDef =   new DefaultTransactionDefinition()
+ITransactionManager tm = odiInstance.getTransactionManager()
+IOdiEntityManager tem = odiInstance.getTransactionalEntityManager()
+ITransactionStatus txnStatus = tm.getTransaction(txnDef)
```
[the remark this answers is not in the document]
Comment (jodi-one) [reply jodi_odi12/src/main/groovy/findLoadPlansLastModified.groovy:19]: It opens a DB transaction  it i is required see; https://blogs.oracle.com/dataintegration/post/odi-12c-mapping-sdk-the-ins-and-outs

File: jodi_odi12/src/main/groovy/findMappings.groovy [hunk 42]
```diff
@@ -0,0 +1,36 @@
+import bsh.StringUtil
+import jnr.ffi.Struct
+import oracle.odi.core.OdiInstance
```
[the remark this answers is not in the document]
Comment (jodi-one) [reply jodi_odi12/src/main/groovy/findMappings.groovy:3]: Nope OdiInstance is used.

**Pull request merged**